### PR TITLE
Performance: Prefetch repository metadata and blob contents from search results (Experimental)

### PR DESCRIPTION
**Pull request by umpox**
> **Note**
> This PR is only partially complete. This is an experimental change that I have put together after thinking about this.
> Mainly looking for reviews on this as a concept!

## Description

This PR adds preloading support to the search results pages:

### Preload 1: Repository + revision metadata on scroll into view

With this change, we would preload this metadata when a result file|commit|repo is scrolled into view in the search results.

**Why?**:
This query is required before we render **anything** on the repo/file pages. It typically takes `150ms` minimum (generally all our GQL queries do), and noticeably slows down performance.

**Risks of overfetching?**:
- Generally I would expect that most searches will return a smaller set of repository + revisions than actual results, which would be typical of an actual customer deployment (many results, fewer repositories).
- We only fetch when the result is scrolled into view, so we wouldn't expect too many queries at once here.
- We would likely see a larger volume of revision fetches on diff search, but I still wouldn't expect this to cause other performance issues, due to the reasons above and the fact that this query is generally quite small and fast.
- Clicking a result will reuse the memoized observables even if they haven't resolved yet, so we wouldn't ever end up in a situation where we fetch the same data twice.

### Preload 2: Plaintext file blob on hover of `FileMatchChildren`

With this change, we would preload a **plaintext** blob whenever this component is hovered.

**Why?**:
This is the final query that takes a noticeable amount of time before a user actually gets to see the result of their intended action (clicking a file result). The requests can typically range from 150ms-1000ms. 

**Risks of overfetching?**:
- We only start fetching when a user hovers over the actual file match example in the results. We immediately stop fetching when a user hovers away from the specific result.
- We only fetch plaintext blob information here, to mitigate any risks of fetching data that we won't need. Syntax highlighting is skipped, and only loaded when a file is actually clicked on.
- Clicking a result will reuse the memoized observables even if they haven't resolved yet, so we wouldn't ever end up in a situation where we fetch the same data twice.

#### Before (Note how 2 and 3 results are quicker as the repo + revision fetch has already loaded)
https://user-images.githubusercontent.com/9516420/183442946-1e1a4a4a-6f45-4e56-80a3-1e6840e4ecdb.mp4

#### After
https://user-images.githubusercontent.com/9516420/183443742-9b460008-abe3-4de5-ab06-be00a5d8f1aa.mp4

## Test plan

Still WIP, only happy path tested currently. Still need to test edge cases.

## App preview:

- [Web](https://sg-web-tr-prefetch-from-search.onrender.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/shared/src/components/PreloadableFile.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    return (
+        <Component
+            onMouseOver={preload ? startPreload : undefined}
+            onMouseLeave={preload ? stopPreload : undefined}
+            ref={reference}
+            {...rest}
+        />
+    )
+}) as ForwardReferenceComponent<'div', PreloadableFileProps>
```
Comment (umpox) [inline client/shared/src/components/PreloadableFile.tsx:73]: This is where the file prefetch start/stop logic happens

File: client/shared/src/backend/repo.ts [hunk 2]
```diff
@@ -102,3 +110,147 @@ export const fetchTreeEntries = memoizeObservable(
...
+            displayName
+            abbrevName
+        }
+    }
+`
+
+/**
+ * Fetch the repository.
+ */
+export const fetchRepository = memoizeObservable(
```
Comment (umpox) [inline client/shared/src/backend/repo.ts:135]: These functions have been moved over from `web` to `shared` so they can also be used in `search-ui`. Need to check how this would integrate with other views (e.g. vscode), or if we don't want this at all

File: client/search-ui/src/components/ResultContainer.tsx [hunk 3]
```diff
@@ -131,20 +140,64 @@ export const ResultContainer: React.FunctionComponent<React.PropsWithChildren<Re
...
+            repoObservable = fetchRepository({
+                repoName,
+                requestGraphQL: platformContext.requestGraphQL,
+            }).subscribe()
+            revisionObservable = resolveRevision({
+                repoName,
+                revision: repoRevision,
+                requestGraphQL: platformContext.requestGraphQL,
+            }).subscribe()
+        }
```
Comment (umpox) [inline client/search-ui/src/components/ResultContainer.tsx:193]: This is where the repository metadata + revision preload happens. This only runs when this result becomes visible, as it's [wrapped in a VirtualList](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@tr/prefetch-from-search/-/blob/client/search-ui/src/results/StreamingSearchResultsList.tsx?L190%3A1-191%3A1=)

File: client/search-ui/src/components/FileMatchChildren.tsx [hunk 4]
```diff
@@ -149,6 +152,13 @@ function navigateToFileOnMiddleMouseButtonClick(event: MouseEvent<HTMLElement>):
 
 export const FileMatchChildren: React.FunctionComponent<React.PropsWithChildren<FileMatchProps>> = props => {
     const [coreWorkflowImprovementsEnabled] = useCoreWorkflowImprovementsEnabled()
+    const prefetchFileEnabled = useMemo(
+        () =>
+            isSettingsValid(props.settingsCascade) &&
+            props.settingsCascade.final.experimentalFeatures?.enableLazyBlobSyntaxHighlighting,
```
Comment (umpox) [inline client/search-ui/src/components/FileMatchChildren.tsx:158]: Currently using `enableLazyBlobSyntaxHighlighting`. Need to update to be its own flag, currently helps with testing against deployed envs though.

**Comment by umpox** [comment]
@fkling @limitedmage Would be interested to get your 👀 on this as a concept and experimental feature.

TLDR: 
- This PR means we preload `RepositoryRedirect` + `ResolveRev` queries for visible results after search submission.
- It also means we preload plaintext `Blob` queries on hover. We cancel the preloading when hover is removed.
- Uses the existing memoized observables so data won't be fetched twice, and can just be resumed by the component where it is used until it has resolved.

This is quite an aggressive prefetch strategy, there's areas where I'm still not fully sure:
- Do we feel this will harm the UX in any way? We will inevitably over fetch some data here. If we were to enable this fully, I'd want it to always be configurable to be disabled for whatever reason (e.g. data usage). I would guess there would be a small additional CPU overhead, but that is likely very small and wouldn't impact our typical users. We also have areas where we send a lot more network requests (e.g. symbol hovers) and I don't think we've ran into many issues related to that (?).
- Is our backend setup to cope with the additional load? I asked [here](https://sourcegraph.slack.com/archives/CHXHX7XAS/p1659615910195859). My main concern would be the `Blob` query, but I think that would be less of an issue now we can skip syntax highlighting.
- What about other search-ui clients, e.g. VSCode? Haven't looked further into these tbh, I would say this would be a 4.0 only change and we would disable on other clients. I think they have a different blob fetch strategy anyway, but unsure.
- Is this worth doing, we're already quite fast? I would still say yes. Search result->View code is arguably the most common workflow on Sourcegraph, and anything that noticeably speeds it up is worth doing IMO.

**Comment by umpox** [comment]
Note: Closing in favour of https://github.com/sourcegraph/sourcegraph/pull/40358 and https://github.com/sourcegraph/sourcegraph/pull/40244 and https://github.com/sourcegraph/sourcegraph/pull/40354
